proto/src/protobuf/mod.rs

```diff
@@ -0,0 +1,215 @@
+mod erased;
+mod error;
+
+use alloc::vec::Vec;
+use core::fmt::Display;
+
+use bytes::Buf;
+use prost::encoding::encoded_len_varint;
+use prost::Message;
+
+pub use self::error::Error;
+
+/// Object safe equivalent of `tendermint_proto::Protobuf`.
+///
+/// Allows for easy Google Protocol Buffers encoding and decoding of domain
+/// types with validation.
+///
+/// Domain types implementing this trait are expected to implement `TryFrom<T>` and `Into<T>` where
+/// `T` is the raw type. The equivalent object safe `erased` counterparts have blanket
+/// implementations and are derived automatically.
+///
+/// ## Examples
+///
+/// ```rust
+/// use core::convert::TryFrom;
+///
+/// use prost::Message;
+/// use ibc_proto::protobuf::Protobuf;
+///
+/// // This struct would ordinarily be automatically generated by prost.
+/// #[derive(Clone, PartialEq, Message)]
+/// pub struct MyRawType {
+///     #[prost(uint64, tag="1")]
+///     pub a: u64,
+///     #[prost(string, tag="2")]
+///     pub b: String,
+/// }
+///
+/// #[derive(Clone)]
+/// pub struct MyDomainType {
+///     a: u64,
+///     b: String,
+/// }
+///
+/// impl MyDomainType {
+///     /// Trivial constructor with basic validation logic.
+///     pub fn new(a: u64, b: String) -> Result<Self, String> {
+///         if a < 1 {
+///             return Err("a must be greater than 0".to_owned());
+///         }
+///         Ok(Self { a, b })
+///     }
+/// }
+///
+/// impl TryFrom<MyRawType> for MyDomainType {
+///     type Error = String;
+///
+///     fn try_from(value: MyRawType) -> Result<Self, Self::Error> {
+///         Self::new(value.a, value.b)
+///     }
+/// }
+///
+/// impl From<MyDomainType> for MyRawType {
+///     fn from(value: MyDomainType) -> Self {
+///         Self { a: value.a, b: value.b }
+///     }
+/// }
+///
+/// impl Protobuf<MyRawType> for MyDomainType {}
+///
```

Comment:
These comments seem to document an example of what this module enables, and don't necessarily only pertain to the `Protobuf` trait. As a result, I wonder if these comments would be more appropriate as module-level docs using `//!` at the top of the file. What do you think?

Reply:
Good point! Done -> https://github.com/informalsystems/ibc-rs/pull/2412/commits/8de194b12486bcdbdba981cca6b93afca4c4dfab
PS: the doc comments were copied from `tendermint-rs` so credits to them for the thorough documentation. :tada: :sweat_smile: